conn.go

```diff
@@ -1326,6 +1326,7 @@ func (c *Conn) deallocateInvalidatedCachedStatements(ctx context.Context) error
 
 	for _, sd := range invalidatedStatements {
 		pipeline.SendDeallocate(sd.Name)
+		delete(c.preparedStatements, sd.Name)
```

Comment:
this actually might not solve the "leak" issue, since Go won't deallocate memory after you delete a key from a map https://github.com/golang/go/issues/20135

Reply:
The map *itself* won't shrink, but `delete()` does zero the keys and values in the map, so the `*pgconn.StatementDescription` value will be garbage collected. Additionally, `delete()` decrements the size of the map, which means its capacity will now be limited to the max number of prepared statements (with extra space for the map "fill factor"). This means this change should mean the amount of memory used by the c.preparedStatements is now fixed, rather than increasing without bound.

Additionally, my reproduction program on the original issue can run forever without running out of memory: https://github.com/jackc/pgx/issues/1456#issuecomment-1550373854

So: I'm pretty confident that this fix will resolve this specific bug at least. There could be others though! I would be happy to try to figure them out if we have a sample program that reproduces the problem.